lib/rubygems/version.rb

```diff
@@ -420,13 +420,23 @@ def freeze
   protected
 
   def _segments
+    return @segments if @segments
     # segments is lazy so it can pick up version values that come from
     # old marshaled versions, which don't go through marshal_load.
     # since this version object is cached in @@all, its @segments should be frozen
 
-    @segments ||= @version.scan(/[0-9]+|[a-z]+/i).map! do |s|
-      /^\d+$/.match?(s) ? s.to_i : -s
-    end.freeze
+    raise "How is this not nil? #{@segments.inspect}" unless @segments.nil?
+
+    @segments = Array.new
+
+    raise "How is a new array frozen? #{@segments.inspect}" if @segments.frozen?
+
+    @version.scan(/\d+|[a-z]+/i) do |s|
+      raise "How did this become frozen? #{@segments.inspect}" if @segments.frozen?
+
+      @segments << (/^\d+$/.match?(s) ? s.to_i : -s)
```

Comment:
Ok, riddle me this. How is this printing the error: 

```
/home/runner/.rubies/truffleruby-22.3.0/lib/ruby/site_ruby/3.0.3.22.3.0/rubygems/version.rb:437:in `<<': can't modify frozen Array (FrozenError)
```

on both [jruby](https://github.com/rubygems/rubygems/actions/runs/6486538501/job/17614928473?pr=7051#step:11:12) and [truffleruby](https://github.com/rubygems/rubygems/actions/runs/6486538501/job/17614928971?pr=7051#step:11:33)

I checked if it's frozen before each iteration. What Array could possibly be frozen?

Reply:
I seed that truffleruby has Gem::Version in it, but is it using this code? It can't be using both.

https://github.com/oracle/truffleruby/blob/21ad763c46422459aba15838b0168d89d91c1bef/lib/mri/rubygems/version.rb#L230

Reply:
what happens if you assign the new array to a local variable instead of of the ivar, and only assign to the ivar after freezing? im wondering if it could be concurrent modification of the version instance

Reply:
Good idea. That would also make sense about why it's intermittent. Seems this class should be atomic in its modifications, or immutable. (I did run a test with freezing on create, it does allocate a little more memory, but it's not dramatic).

Reply:
This is the branch where I immediately freeze at the end of initialize compared to the exact same code not doing that:

```
==> memprof.nofreeze.txt <==
Total allocated: 24.85 MB (206507 objects)
Total retained:  3.71 MB (38062 objects)
head: : No such file or directory
==> memprof.2.4.20.txt <==
Total allocated: 28.57 MB (255984 objects)
Total retained:  4.09 MB (42873 objects)
==> memprof.frozen.txt <==
Total allocated: 24.95 MB (207562 objects)
Total retained:  3.72 MB (38216 objects)
```